Build failure with PG11: error: macro "AllocSetContextCreate" passed 5 arguments, but takes just 3

pllua fails to build against PG 11beta1:
```
11:17:41 gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -g -g -O2 -fdebug-prefix-map=/build/postgresql-11-LcCi7f/postgresql-11-11~beta1=. -fstack-protector-strong -Wformat -Werror=format-security -fno-omit-frame-pointer -fPIC -I/usr/include/lua5.1  -I. -I/<<PKGBUILDDIR>> -I/usr/include/postgresql/11/server -I/usr/include/postgresql/internal  -Wdate-time -D_FORTIFY_SOURCE=2 -D_GNU_SOURCE -I/usr/include/libxml2  -I/usr/include/mit-krb5  -c -o pllua_xact_cleanup.o /<<PKGBUILDDIR>>/pllua_xact_cleanup.c
11:17:41 /<<PKGBUILDDIR>>/pllua_xact_cleanup.c: In function 'pllua_init_common_ctx':
11:17:41 /<<PKGBUILDDIR>>/pllua_xact_cleanup.c:136:33: error: macro "AllocSetContextCreate" passed 5 arguments, but takes just 3
11:17:41          ALLOCSET_DEFAULT_MAXSIZE);
11:17:41                                  ^
11:17:41 /<<PKGBUILDDIR>>/pllua_xact_cleanup.c:134:15: error: 'AllocSetContextCreate' undeclared (first use in this function); did you mean 'SlabContextCreate'?
11:17:41      cmn_ctx = AllocSetContextCreate(TopMemoryContext,
11:17:41                ^~~~~~~~~~~~~~~~~~~~~
11:17:41                SlabContextCreate
11:17:41 /<<PKGBUILDDIR>>/pllua_xact_cleanup.c:134:15: note: each undeclared identifier is reported only once for each function it appears in
11:17:42 make[2]: *** [<builtin>: pllua_xact_cleanup.o] Error 1
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -51,6 +51,9 @@ env:
     - PGVERSION=10 LUA=lua5.1 LUA_DEV=liblua5.1-dev LUA_INCDIR=/usr/include/lua5.1 LUALIB=-llua5.1
     - PGVERSION=10 LUA=lua5.3 LUA_DEV=liblua5.3-dev LUA_INCDIR=/usr/include/lua5.3 LUALIB=-llua5.3
     - PGVERSION=10 LUA=luajit LUA_DEV=libluajit-5.1-dev LUA_INCDIR=/usr/include/luajit-2.0 LUALIB=-lluajit-5.1
+    - PGVERSION=11 LUA=lua5.1 LUA_DEV=liblua5.1-dev LUA_INCDIR=/usr/include/lua5.1 LUALIB=-llua5.1
+    - PGVERSION=11 LUA=lua5.3 LUA_DEV=liblua5.3-dev LUA_INCDIR=/usr/include/lua5.3 LUALIB=-llua5.3
+    - PGVERSION=11 LUA=luajit LUA_DEV=libluajit-5.1-dev LUA_INCDIR=/usr/include/luajit-2.0 LUALIB=-lluajit-5.1
 
 
 language: c

--- a/pllua_pgfunc.c
+++ b/pllua_pgfunc.c
@@ -109,11 +109,7 @@ static MemoryContext
 get_tmpcontext()
 {
 	MemoryContext mc;
-	mc = AllocSetContextCreate(TopMemoryContext,
-				   "pgfunc temporary context",
-				   ALLOCSET_DEFAULT_MINSIZE,
-				   ALLOCSET_DEFAULT_INITSIZE,
-				   ALLOCSET_DEFAULT_MAXSIZE);
+	mc = pg_create_context("pgfunc temporary context");
 	return mc;
 }
 #endif

--- a/pllua_xact_cleanup.c
+++ b/pllua_xact_cleanup.c
@@ -131,9 +131,7 @@ void pllua_xact_cb(XactEvent event, void *arg)
 
 void pllua_init_common_ctx()
 {
-    cmn_ctx = AllocSetContextCreate(TopMemoryContext,
-        "PL/Lua common context", ALLOCSET_DEFAULT_MINSIZE, ALLOCSET_DEFAULT_INITSIZE,
-        ALLOCSET_DEFAULT_MAXSIZE);
+    cmn_ctx = pg_create_context("PL/Lua common context");
 
     resource_stk = rsp_initStack();
 }

--- a/plluaapi.c
+++ b/plluaapi.c
@@ -510,9 +510,7 @@ lua_State *luaP_newstate (int trusted) {
   int status;
 
 
-  MemoryContext mcxt = AllocSetContextCreate(TopMemoryContext,
-      "PL/Lua context", ALLOCSET_DEFAULT_MINSIZE, ALLOCSET_DEFAULT_INITSIZE,
-      ALLOCSET_DEFAULT_MAXSIZE);
+  MemoryContext mcxt = pg_create_context("PL/Lua context");
   lua_State *L = luaL_newstate();
   lua_atpanic(L, luaP_panic);
   /* version */
@@ -906,11 +904,10 @@ void luaP_pushdatum (lua_State *L, Datum dat, Oid type) {
           Datum value;
           lua_createtable(L, 0, ti->tupdesc->natts);
           for (i = 0; i < ti->tupdesc->natts; i++) {
-            Form_pg_attribute att = ti->tupdesc->attrs[i];
-            key = NameStr(att->attname);
-            value = GetAttributeByNum(tup, att->attnum, &isnull);
+            key = NameStr(TupleDescAttr(ti->tupdesc, i)->attname);
+            value = GetAttributeByNum(tup, TupleDescAttr(ti->tupdesc, i)->attnum, &isnull);
             if (!isnull) {
-              luaP_pushdatum(L, value, att->atttypid);
+              luaP_pushdatum(L, value, TupleDescAttr(ti->tupdesc, i)->atttypid);
               lua_setfield(L, -2, key);
             }
           }
@@ -1140,10 +1137,11 @@ Datum luaP_todatum (lua_State *L, Oid type, int typmod, bool *isnull, int idx) {
               /* create tuple */
               b = luaP_getbuffer(L, ti->tupdesc->natts);
               for (i = 0; i < ti->tupdesc->natts; i++) {
-                lua_getfield(L, idx, NameStr(ti->tupdesc->attrs[i]->attname));
+
+                lua_getfield(L, idx, NameStr(TupleDescAttr(ti->tupdesc, i)->attname));
                 /* only simple types allowed in record */
-                b->value[i] = luaP_todatum(L, ti->tupdesc->attrs[i]->atttypid,
-                    ti->tupdesc->attrs[i]->atttypmod, b->null + i, idx);
+                b->value[i] = luaP_todatum(L, TupleDescAttr(ti->tupdesc, i)->atttypid,
+                    TupleDescAttr(ti->tupdesc, i)->atttypmod, b->null + i, idx);
                 lua_pop(L, 1);
               }
               /* make copy in upper executor memory context */

--- a/plluacommon.h
+++ b/plluacommon.h
@@ -82,6 +82,18 @@ MemoryContext luaP_getmemctxt (lua_State *L);
 lua_State *pllua_getmaster (lua_State *L);
 int pllua_getmaster_index(lua_State *L);
 
+#if PG_VERSION_NUM < 110000
+    #define TupleDescAttr(tupdesc, i) ((tupdesc)->attrs[(i)])
+
+    #define pg_create_context(name) \
+            AllocSetContextCreate(TopMemoryContext, \
+                name, ALLOCSET_DEFAULT_MINSIZE, ALLOCSET_DEFAULT_INITSIZE, \
+                ALLOCSET_DEFAULT_MAXSIZE);
+#else
+    #define pg_create_context(name) \
+            AllocSetContextCreate(TopMemoryContext, name, ALLOCSET_DEFAULT_SIZES);
+#endif
+
 #define lua_swap(L) lua_insert(L, -2)
 
 #define luaP_getfield(L, s) \

--- a/plluaspi.c
+++ b/plluaspi.c
@@ -101,7 +101,7 @@ void luaP_pushdesctable (lua_State *L, TupleDesc desc) {
   int i;
   lua_newtable(L);
   for (i = 0; i < desc->natts; i++) {
-    lua_pushstring(L, NameStr(desc->attrs[i]->attname));
+    lua_pushstring(L, NameStr(TupleDescAttr(desc, i)->attname));
     lua_pushinteger(L, i);
     lua_rawset(L, -3); /* t[att] = i */
   }
@@ -126,7 +126,7 @@ static void luaP_pushtuple_cmn (lua_State *L, HeapTuple tuple,
     for (i = 0; i < n; i++) {
 
         bool isnull;
-        t->value[i] = heap_getattr(tuple, tupleDesc->attrs[i]->attnum, tupleDesc,
+        t->value[i] = heap_getattr(tuple, TupleDescAttr(tupleDesc, i)->attnum, tupleDesc,
                                    &isnull);
         t->null[i] = isnull;
     }
@@ -376,7 +376,7 @@ static luaP_Tuple* luaP_PTuple_rawctr(lua_State * L, HeapTuple tuple, int readon
     t->rtupdesc = rtupdesc_ref(rtupdesc);
     for (i = 0; i < n; i++) {
         bool isnull;
-        t->value[i] = heap_getattr(tuple, tupleDesc->attrs[i]->attnum, tupleDesc,
+        t->value[i] = heap_getattr(tuple, TupleDescAttr(tupleDesc, i)->attnum, tupleDesc,
                                    &isnull);
         t->null[i] = isnull;
     }
@@ -435,7 +435,7 @@ static int luaP_p_tupleindex (lua_State *L) {
             }
             if ((i >= 0)&&(i < tupleDesc->natts)) {
                 if (!t->null[i])
-                    luaP_pushdatum(L, t->value[i], tupleDesc->attrs[i]->atttypid);
+                    luaP_pushdatum(L, t->value[i], TupleDescAttr(tupleDesc, i)->atttypid);
                 else lua_pushnil(L);
             }
             else {
@@ -460,7 +460,7 @@ static int luaP_p_tupleindex (lua_State *L) {
         }
         for (i = 0; i< tupleDesc->natts; ++i){
 
-            if (strcmp(NameStr(tupleDesc->attrs[i]->attname),name) == 0){
+            if (strcmp(NameStr(TupleDescAttr(tupleDesc, i)->attname),name) == 0){
                 idx = i;
                 break;
             }
@@ -468,7 +468,7 @@ static int luaP_p_tupleindex (lua_State *L) {
         i = idx;
         if (i >= 0) {
             if (!t->null[i])
-                luaP_pushdatum(L, t->value[i], tupleDesc->attrs[i]->atttypid);
+                luaP_pushdatum(L, t->value[i], TupleDescAttr(tupleDesc, i)->atttypid);
             else lua_pushnil(L);
         }
         else {
@@ -497,7 +497,7 @@ static int luaP_tupleindex (lua_State *L) {
       }
       for (i = 0; i< tupleDesc->natts; ++i){
 
-          if (strcmp(NameStr(tupleDesc->attrs[i]->attname),name) == 0){
+          if (strcmp(NameStr(TupleDescAttr(tupleDesc, i)->attname),name) == 0){
               idx = i;
               break;
           }
@@ -505,7 +505,7 @@ static int luaP_tupleindex (lua_State *L) {
       i = idx;
       if (i >= 0) {
           if (!t->null[i])
-            luaP_pushdatum(L, t->value[i], tupleDesc->attrs[i]->atttypid);
+            luaP_pushdatum(L, t->value[i], TupleDescAttr(tupleDesc, i)->atttypid);
           else lua_pushnil(L);
       }
       else {
@@ -525,7 +525,7 @@ static int luaP_tupleindex (lua_State *L) {
 
   if (i >= 0) {
       if (!t->null[i])
-        luaP_pushdatum(L, t->value[i], t->tupdesc->attrs[i]->atttypid);
+        luaP_pushdatum(L, t->value[i], TupleDescAttr(t->tupdesc, i)->atttypid);
       else lua_pushnil(L);
   }
   else lua_pushnil(L);
@@ -545,8 +545,8 @@ static int luaP_tuplenewindex (lua_State *L) {
   lua_settop(L, 3);
   if (i >= 0) { /* found? */
     bool isnull;
-    t->value[i] = luaP_todatum(L, t->tupdesc->attrs[i]->atttypid,
-        t->tupdesc->attrs[i]->atttypmod, &isnull, -1);
+    t->value[i] = luaP_todatum(L, TupleDescAttr(t->tupdesc, i)->atttypid,
+        TupleDescAttr(t->tupdesc, i)->atttypmod, &isnull, -1);
     t->null[i] = isnull;
     t->changed = 1;
   }
@@ -584,7 +584,7 @@ void luaP_pushtuple_trg (lua_State *L, TupleDesc desc, HeapTuple tuple,
     t->rtupdesc = 0;
     for (i = 0; i < n; i++) {
         bool isnull;
-        t->value[i] = heap_getattr(tuple, desc->attrs[i]->attnum, desc,
+        t->value[i] = heap_getattr(tuple, TupleDescAttr(desc, i)->attnum, desc,
                                    &isnull);
         t->null[i] = isnull;
     }
@@ -641,12 +641,12 @@ HeapTuple luaP_casttuple (lua_State *L, TupleDesc tupdesc) {
   b = luaP_getbuffer(L, tupdesc->natts);
   for (i = 0; i < tupdesc->natts; i++) {
     int j;
-    lua_getfield(L, -1, NameStr(tupdesc->attrs[i]->attname));
+    lua_getfield(L, -1, NameStr(TupleDescAttr(tupdesc, i)->attname));
     j = luaL_optinteger(L, -1, -1);
     if (j >= 0) {
       if (t->changed == -1) /* read-only? */
         b->value[i] = heap_getattr(t->tuple,
-            t->tupdesc->attrs[j]->attnum, t->tupdesc, b->null + i);
+            TupleDescAttr(t->tupdesc, j)->attnum, t->tupdesc, b->null + i);
       else {
         b->value[i] = t->value[j];
         b->null[i] = t->null[j];

